test/integration/statefulset/util.go

```diff
@@ -0,0 +1,329 @@
+func scSetup(t *testing.T) (*httptest.Server, framework.CloseFunc, *statefulset.StatefulSetController, informers.SharedInformerFactory, clientset.Interface) {
+	masterConfig := framework.NewIntegrationTestMasterConfig()
+	_, s, closeFn := framework.RunAMaster(masterConfig)
+
+	config := restclient.Config{Host: s.URL}
+	clientSet, err := clientset.NewForConfig(&config)
+	if err != nil {
+		t.Fatalf("error in create clientset: %v", err)
+	}
+	resyncPeriod := 12 * time.Hour
+	informers := informers.NewSharedInformerFactory(clientset.NewForConfigOrDie(restclient.AddUserAgent(&config, "statefulset-informers")), resyncPeriod)
+
+	sc := statefulset.NewStatefulSetController(
+		informers.Core().V1().Pods(),
+		informers.Apps().V1().StatefulSets(),
+		informers.Core().V1().PersistentVolumeClaims(),
+		informers.Apps().V1().ControllerRevisions(),
+		clientset.NewForConfigOrDie(restclient.AddUserAgent(&config, "statefulset-controller")),
+	)
+
+	return s, closeFn, sc, informers, clientSet
+}
+
+// Run STS controller and informers
+func runControllerAndInformers(sc *statefulset.StatefulSetController, informers informers.SharedInformerFactory) chan struct{} {
+	stopCh := make(chan struct{})
+	informers.Start(stopCh)
+	go sc.Run(5, stopCh)
+	return stopCh
+}
+
+func createHeadlessService(t *testing.T, clientSet clientset.Interface, headlessService *v1.Service) {
+	_, err := clientSet.Core().Services(headlessService.Namespace).Create(headlessService)
+	if err != nil {
+		t.Fatalf("Failed creating headless service: %v", err)
+	}
+}
+
+func createSTSsPods(t *testing.T, clientSet clientset.Interface, stss []*v1beta1.StatefulSet, pods []*v1.Pod) ([]*v1beta1.StatefulSet, []*v1.Pod) {
+	var ns string
+	if len(stss) != 0 {
+		ns = stss[0].Namespace
+	} else {
+		ns = pods[0].Namespace
+	}
+	stsClient := clientSet.AppsV1beta1().StatefulSets(ns)
+	podClient := clientSet.Core().Pods(ns)
+	var createdSTSs []*v1beta1.StatefulSet
+	var createdPods []*v1.Pod
+	for _, sts := range stss {
+		createdSTS, err := stsClient.Create(sts)
+		if err != nil {
+			t.Fatalf("Failed to create sts %s: %v", sts.Name, err)
+		}
+		createdSTSs = append(createdSTSs, createdSTS)
+	}
+	for _, pod := range pods {
+		createdPod, err := podClient.Create(pod)
+		if err != nil {
+			t.Fatalf("Failed to create pod %s: %v", pod.Name, err)
+		}
+		createdPods = append(createdPods, createdPod)
+	}
+
+	return createdSTSs, createdPods
+}
+
+// Verify .Status.Replicas is equal to .Spec.Replicas
+func waitSTSStable(t *testing.T, clientSet clientset.Interface, sts *v1beta1.StatefulSet) {
+	stsClient := clientSet.AppsV1beta1().StatefulSets(sts.Namespace)
+	if err := wait.PollImmediate(interval, timeout, func() (bool, error) {
+		newSTS, err := stsClient.Get(sts.Name, metav1.GetOptions{})
+		if err != nil {
+			return false, err
+		}
+		return newSTS.Status.Replicas == *newSTS.Spec.Replicas, nil
```

Comment:
We should also verify `.Status.ObservedGeneration >= desiredGeneration`. Refer this: https://github.com/kubernetes/kubernetes/blob/master/test/utils/replicaset.go#L56-L68